packages/app-lib/src/util/rpc.rs

```diff
@@ -0,0 +1,233 @@
+            )
+            .into());
+        }
+
+        tracing::debug!("Waiting on result for {id}");
+        let Ok(result) = recv.await else {
+            self.waiting_responses.lock().unwrap().remove(&id);
+            return Err(ErrorKind::RpcError(
+                "RPC connection closed while waiting for response".to_string(),
+            )
+            .into());
+        };
+        result.and_then(|x| Ok(serde_json::from_value(x)?))
+    }
+
+    async fn run(
+        socket: TcpListener,
+        mut message_receiver: mpsc::UnboundedReceiver<RpcMessage>,
+        handlers: HandlerMap,
+        waiting_responses: WaitingResponsesMap,
+    ) -> Result<()> {
+        let (mut socket, _) = socket.accept().await?;
+        loop {
+            tokio::select! {
+                message = message_receiver.recv() => {
+                    let Some(message) = message else {
+                        break;
+                    };
+                    tracing::debug!("Sending RPC message {message:?}");
+                    let json = serde_json::to_vec(&message)?;
+                    socket.write_all(&json).await?;
+                    socket.flush().await?;
+                },
+                message = parse_object_async_reader(&mut socket) => {
+                    let message: RpcMessage = message?;
+                    tracing::debug!("Received RPC message {message:?}");
+                    if let RpcMessageBody::Call { method, args } = message.body {
+                        let response = match handlers.get(method.as_str()) {
+                            Some(handler) => match handler(args).await {
+                                Ok(result) => RpcMessageBody::Respond {
+                                    response: result,
+                                },
+                                Err(e) => RpcMessageBody::Error {
+                                    error: e.to_string(),
+                                },
+                            }
+                            None => RpcMessageBody::Error {
+                                error: format!("Unknown theseus RPC method {method}"),
+                            },
+                        };
+                        let json = serde_json::to_vec(&RpcMessage {
+                            id: message.id,
+                            body: response,
+                        })?;
+                        socket.write_all(&json).await?;
+                        socket.flush().await?;
+                    } else if let Some(sender) = waiting_responses.lock().unwrap().remove(&message.id) {
+                        let _ = sender.send(match message.body {
+                            RpcMessageBody::Respond { response } => Ok(response),
+                            RpcMessageBody::Error { error } => Err(ErrorKind::RpcError(error).into()),
+                            _ => unreachable!(),
+                        });
+                    }
+                },
```

Comment:
Out of curiosity, do you get good `rust-analyzer`/IDE diagnostics out of code within a `select!` macro? I don't, so I default to putting as little code in them as possible, but maybe there is a way to fix that :thinking: